JBMC: enable model and SSA equation verification

smowton opened the pull request:
This enables validation of the GOTO model and SSA equations in all JBMC regression tests, making the minimal modifications such that they work:

* Fixing an accidentally untyped expression
* Disabling basename validation for Java, which presently uses a different convention that C for name-to-basename relationship
* Loosening type comparison to only require `base_type_eq`, not exact equality
* Fixing SSA expression type verification
* Fixing SSA DECL verification

I package these together here because this enables all these small fixes to be checked throughout the whole regression test suite, but if people prefer they can be split apart. Naturally, recommend reviewing commit by commit.

smowton committed:
Method stubs: add missing type

This had used a typeless symbol expression for ages without us noticing.

smowton committed:
JBMC: support validation

This invokes validation using the same arguments as CBMC.

smowton committed:
Lazy goto model: support validation

This mirrors existing support in goto_modelt.

smowton committed:
Don't validate Java symbols' base names for now

They use a different pattern to that expected for C programs, and changing that will have some impact. For now
let's make validation practically usable without making a change with potentially widespread impact.

smowton committed:
Enable model validation in all JBMC regression tests

smowton committed:
DECL VCCs: don't check non-existent SSA-rhs

DECL clauses do not give an RHS, so comparing its type against the LHS is meaningless (and always fails).

smowton committed:
Equation validation: use base_type_eq on types, not exprs

The expr version expects two like-structured expressions and validates subexpressions pairwise,
which is not the expected situation for VCCs.

smowton committed:
JBMC: enable SSA equation validation in all regression tests

tautschnig commented on src/util/symbol.cpp:
```diff
@@ -137,7 +137,12 @@ bool symbolt::is_well_formed() const
+  // "method")
+  if(
+    !has_suffix(id2string(name), id2string(base_name)) &&
+    mode != ID_java)
```
I believe @chrisr-diffblue was also looking into this as there are also cases in C where this doesn't hold. Is this correct?

chrisr-diffblue replied:
Yes, I'm looking into other cases - just hadn't got as far as java yet, so Chris has saved me some work here :-)

tautschnig commented on jbmc/src/java_bytecode/simple_method_stubbing.cpp:
```diff
@@ -170,7 +170,8 @@ void java_simple_method_stubst::create_method_stub(symbolt &symbol)
     code_assignt get_argument(
-      init_symbol_expression, symbol_exprt(this_argument.get_identifier()));
+      init_symbol_expression,
+      symbol_exprt(this_argument.get_identifier(), this_type));
```
We should really remove the single-argument constructor.

smowton replied:
In another PR, sure

tautschnig replied:
Sure, that was not meant as a request for work.

tautschnig commented on src/util/std_code.h:
```diff
@@ -296,14 +297,14 @@ class code_assignt:public codet
     DATA_CHECK(
       vm,
-      code.op0().type() == code.op1().type(),
+      base_type_eq(code.op0().type(), code.op1().type(), ns),
```
To be debated with @kroening as this appears to work against the effort going on in other PRs.

smowton replied:
Fair enough if we want the strict same-level-of-indirection condition in the future, but we certainly don't even try to do that at present.

tautschnig replied:
Can this commit be dropped from the PR? I understand that this would mean it doesn't work for Java at the moment, but then we can get all the other improvements in the meantime.

smowton replied:
Well, my intent was to use this as the means of testing the other improvements, by showing particular programs begin to satisfy this constraint when the fixes are applied, so I'd much prefer to have this statement of our current intended guarantee available, then we can chase after a future attempted standard later?

tautschnig commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -966,6 +966,8 @@ void symex_target_equationt::SSA_stept::validate(
     validate_full_expr(ssa_full_lhs, ns, vm);
     validate_full_expr(original_full_lhs, ns, vm);
+    if(type == goto_trace_stept::typet::DECL)
+      break;
```
In this case I'd prefer some copy&paste and just completely separate the `ASSIGNMENT` and `DECL` cases.

chrisr-diffblue replied:
+1

smowton replied:
done

chrisr-diffblue reviewed (approved):
Not going to comment on the `base_type_eq` debate, but I'm happy for this to go in as it is. I'll leave the final say on `base_type_eq` to @tautschnig and @kroening

smowton commented:
@tautschnig spoke with @kroening just now, who said he's happy to go for `base_type_eq` now (and fix everything that violates even *that*) and then subsequently go to `==`.

tautschnig commented:
@smowton Maybe then do a clean revert of the commit that switched from `base_type_eq` to `==` recently to make clear we are intentionally reverting this change?

smowton committed:
Revert "strengthen assignment type consistency"

This reverts commit caae4ceee630400519fd47528dadedc62149c00b.

Reverting because we can't currently meet either the weak (base_type_eq) or the strong (==) consistency requirement.
Let's target the weak one first.

smowton commented:
@tautschnig done

smowton committed:
Pacify clang-format

Adding an include in a prior commit --> I must sort the other ones, apparently :)

smowton committed:
Add name to namespace parameter

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4232b15cc545bca9b426217d60093c6767cf6489).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97290653

peterschrammel commented on src/util/symbol.cpp:
```diff
@@ -137,10 +137,11 @@ bool symbolt::is_well_formed() const
+  // Exception: Java symbols' base names do not have type signatures
+  // (for example, they can have name "someclass.method:(II)V" and base name
+  // "method")
+  if(!has_suffix(id2string(name), id2string(base_name)) && mode != ID_java)
```
Can we check that `method` is between `.` and `:` instead?

smowton replied:
It also differs for field symbols, and generics, and... I'd like to leave (a) setting a convention and (b) enforcing it for a different PR!
